File: Competitive Programming Solutions/DMOPC/dmopc19c3p3.cpp
#include <bits/stdc++.h>
#include <iostream>
using namespace std;
int N,n[150000],fre[300002]={0},mid=150001,shift=150001;
long long total = 0;
int main() {
    cin.sync_with_stdio(0);
    cin.tie(0);
    cin>>N;
	for(int i = 0; i<N; i++) {
		cin>>n[i];
		mid+=n[i]==1?1:-1;
        for(int j = mid; j<=300001; j+=j&-j)fre[j]++;
	}
	long total = 0;
	for(int i = 0; i<N; i++) {
        for(int j = 300001; j>0; j-=j&-j)total+=fre[j];
        for(int j = shift; j>0; j-=j&-j)total-=fre[j];
		shift+=n[i]==1?1:-1;
        for(int j = shift; j<=300001; j+=j&-j)fre[j]--;
	}
	cout<<total<<"\n";
}

File: Programming Solutions/Competitive Programming Solutions/DMOPC/dmopc19c3p3.java
import java.util.*;
import java.io.*;
public class Majority{
	static class BIT{
		long tree[];
		int values[], len;
		public BIT(int length) {
			len=length+1;
			tree=new long[len];
			values=new int[len];
		}
		public void update(int i, int v) {
			for(int j = i; j<len; j+=j&-j)tree[j]+=v;
			values[i]+=v;
		}
		public long query(int i) {
			long sum = 0;
			for(int j = i; j>0; j=j-(j&-j))sum+=tree[j];
			return sum;
		}
	}
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException{
		int N = readInt(), n[] = new int[N], mid = 150001, shift = 150001;
		BIT fre = new BIT(300001);
		for(int i = 0; i<N; i++) {
			n[i]=readInt();
			mid+=n[i]==1?1:-1;
			fre.update(mid, 1);
		}
		long total = 0;
		for(int i = 0; i<N; i++) {
			total+=fre.query(300001)-fre.query(shift);
			shift+=n[i]==1?1:-1;
			fre.update(shift, -1);
		}
		System.out.println(total);
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}